### DISCUSSION TITLE
Limit in region mapping

### DISCUSSION THREAD

**Discussion by josecarlosrodriguezcortes** [TerriaJS/terriajs#5362] [Q&A] [answered]
Hello, I'm having a problem with WMS region mapping.
I have a 250 m2 grid of terrain of data, so I have approximately 120k regions.
My problem is that Terria only draws the first 65535 elements (this is not an arbitrary number, is the last number in the styles that are provided in the docs of Terria and 2^16 - 1) so I have a map with some points drawed but a lot of them not.
I know there is a new region mapping system with vector tiles, but I don't know if this is gonna be a problem with it.
Where could I find the code that checks this so I can find this limit and, idealy, remove it?

![image](https://user-images.githubusercontent.com/73109551/112332493-af1dd500-8cb1-11eb-86ff-cccb43f57ea7.png)

**Comment by tephenavies** [comment, marked as answer]
HI @josecarlosrodriguezcortes. As you've found this limit is the number of defined style colours. The region mapping style uses only blue and green pixel values.

Here is the function that does color -> region FID conversion: https://github.com/TerriaJS/terriajs/blob/6dc8cc397d62de54b52c09bad42965f62db18d06/lib/Map/ImageryProviderHooks.js#L30-L54

I think you could add regions 65536 and onwards using a red value of 1, then 2 (that way it's backwards compatible with the original r = 0 scheme for the first 65535 regions). You'll need to update that function. I think you can do that without forking terriajs given how it's exported. Just write at the top of index.js:
```js
var ImageryProviderHooks = require("terriajs/lib/Map/ImageryProviderHooks.js")

ImageryProviderHooks.recolorImage = function(image, colorFunc) {
  // new function that decodes 0-65535 and 65536+
}
 ```

And then update the geoserver styling to match your extended scheme.

**Reply by tephenavies** [reply]
Also note that WMS region mapping is deprecated and will not be supported in version 8.

**Reply by josecarlosrodriguezcortes** [reply]
Thank you so much, I've been struggling with this for some days.

I know about the new region mapping system using Vector Tiles and I'm planning to move to this version, but this was a necessity just now.
